File: Basis/Blumenerkennung/ErkennungCis/ErkennungCis/Cluster.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Drawing;
using System.Diagnostics;

namespace ErkennungCis
{
    class Cluster
    {

        public static Point[] cluster(Point[] punkte, int dichte, int minpts = 0)
        {
            DateTime start = DateTime.Now;
            int cID = 0;
            Dictionary<Point, int> labels = new Dictionary<Point, int>();
            //Punkte Labeln
            foreach (Point p in punkte)
            {
                if (!labels.ContainsKey(p))
                    labels.Add(p, -2);
            }
            //Lables: -2 = Undefined
            //Lables: -1 = Noise
            //Lables: 0+ = Clusterzug.


            foreach (Point p in punkte)
            {
                if (labels[p] != -2)
                    continue;
                var neighbors = RangeQuery(punkte, p, dichte);
                if (neighbors.Count <= minpts)
                {
                    labels[p] = -1; //Noise
                    continue;
                }
                cID++; //Neues Cluster
                labels[p] = cID; //p neues Cluster
                neighbors.Remove(p);
                for (int i = 0; i < neighbors.Count; i++)
                {
                    Point n = neighbors.ElementAt<Point>(i);
                    if (labels[n] == -1)
                        labels[n] = cID;
                    if (labels[n] != -2)
                        continue;
                    labels[n] = cID;
                    var nn = RangeQuery(punkte, n, dichte);
                    if (nn.Count >= minpts)
                    {
                        mergeLists(neighbors, nn);
                        //Debug.WriteLine("neighbours: " + neighbors.Count+ " CID: "+cID);
                    }
                }

            }

            //Ergebnis fertig machen
            //Pro Cluster ein Punkt
            // Durchschnittlichen Mittelpunkt berechnen
            Point[] punkt = new Point[cID]; //So viele Punkte wie Cluster
            byte[] zähler = new byte[cID]; //o viele Zähler wie Cluster
            //Arrays mit Leerwerten befüllen
            for (int i = 0; i < cID; i++)
            {
                punkt[i] = Point.Empty;
                zähler[i] = 0;
            }
            //Durchschnitt aufaddieren
            foreach (Point p in punkte)
            {
                int lbl = labels[p]-1;
                if (lbl < 0)
                    continue;
                punkt[lbl].X += p.X;
                punkt[lbl].Y += p.Y;
                zähler[lbl]++;

            }

            //Durchschnitt teilen
            for (int i = 0; i < cID; i++)
            {
               punkt[i].X /= zähler[i];
                punkt[i].Y /= zähler[i];
            }
            Debug.WriteLine("{0} Cluster in {1}ms",punkt.Length, DateTime.Now.Subtract(start).TotalMilliseconds);
            return punkt;
        }

        private static LinkedList<Point> RangeQuery(Point[] points, Point mitte, int minD)
        {
            LinkedList<Point> ps = new LinkedList<Point>();
            foreach (Point p in points)
            {
                if (mitte == p)
                    continue;
                if (dist(mitte, p)<=minD)
                {
                    ps.AddFirst(p);
                }
            }
            return ps;
        }

        private static double dist(Point p1, Point p2)
        {
             return Math.Sqrt(Math.Pow(p2.X - p1.X, 2) + Math.Pow(p2.Y - p1.Y, 2));
        }

        private static void mergeLists(LinkedList<Point> main, LinkedList<Point> append)
        {
            foreach (Point p in append)
            {
                main.AddLast(p);
            }
        }


    }
}

File: Drohne/Steuerung/src/bilderkennung/Cluster.java
package bilderkennung;

import java.awt.Point;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.Map;

/*Labels:	-2: Undefined
 * 			-1: Noise
 * 			0+: Cluster ID
 * */

class Cluster {
	
	private Cluster() {}//Privater Konstruktor -> Keine Instanzen
	
	static Point[] cluster(Point[] points) {
		int cId = 0;
		Map<Point, Integer> labels = new HashMap<Point, Integer>();
		// Markiere Punkte
		for (Point p : points) {
			if (!labels.containsKey(p)) {
				labels.put(p, -2);
			}
		}
		
		for (Point p : points) {
			if (labels.get(p) != -2)
				continue;
			
			LinkedList<Point> neighbours = RangeQuery(points, p);
			//Noise Check
			if (neighbours.size() < Einstellungen.minPts) {
				labels.replace(p, -1);
				continue;
			}
			
			cId ++; //Neues Cluster
			labels.replace(p, cId);
			neighbours.remove(p);
			//Expandiere Cluster
			for (int i = 0; i < neighbours.size(); i++) {
				Point n = neighbours.get(i);
				if (labels.get(n) == -1)
					labels.replace(p, cId);
				if (labels.get(n) != -2)
					continue;
				labels.replace(n, cId);
				LinkedList<Point> n2 = RangeQuery(points, n);
				if (n2.size() >= Einstellungen.minPts) {
					mergeLists(neighbours, n2);
				}
			}
			
		}
		
		return durchschnitt(labels, points, cId);
		
	}
	
	private static LinkedList<Point> RangeQuery(Point[] punkte, Point mitte){
		LinkedList<Point> res = new LinkedList<Point>();
		for (Point p : punkte) {
			if (mitte == p)
				continue;
			if (dist(mitte, p) <= Einstellungen.dichte) {
				res.addFirst(p);
			}
		}
		return res;
	}
	
	private static double dist(Point m, Point p) {
		return Math.sqrt(Math.pow(p.x-m.x, 2)+Math.pow(p.y+m.y, 2));
	}
	
	private static void mergeLists(LinkedList<Point> s, LinkedList<Point> d) {
		s.addAll(d);
	}
	
	private static Point[] durchschnitt(Map<Point, Integer> lbl, Point[] pts, int size) {
		Point[] cluster = new Point[size];
		byte[] zähler = new byte[size];
		for (int i = 0; i< size; i++) {
			cluster[i] = new Point(0,0);
			zähler[i] = 0;
		}
		//Jeden Punkt clsuter hinzufügen
		for (Point p : pts) {
			int l = lbl.get(p)-1;
			if (l < 0)
				continue;
			cluster[l].x += p.x;
			cluster[l].y += p.y;
			zähler[l]++;
		}
		//Durchscnitt errechnen
		for (int i = 0; i < cluster.length; i++) {
			cluster[i].x /= zähler[i];
			cluster[i].y /= zähler[i];
		}
		return cluster;
	}
}
